pkg/apis/internal.admin.acorn.io/v1/resources.go

```diff
@@ -7,65 +7,97 @@ import (
+	Jobs       Count = "jobs"
+	Images     Count = "images"
+	Projects   Count = "projects"
+
+	Volumes PersistentCount = "volumes"
+	Secrets PersistentCount = "secrets"
+
+	VolumeStorage Quantity = "volumeStorage"
+	Memory        Quantity = "memory"
+	CPU           Quantity = "cpu"
+)
+
 // Resources is a struct separate from the QuotaRequestInstanceSpec to allow for
 // external controllers to programmatically set the resources easier. Calls to
 // its functions are mutating.
 type Resources struct {
-	Unlimited bool `json:"unlimited,omitempty"`
-
-	Apps       int `json:"apps,omitempty"`
-	Containers int `json:"containers,omitempty"`
-	Jobs       int `json:"jobs,omitempty"`
-	Volumes    int `json:"volumes,omitempty"`
-	Secrets    int `json:"secrets,omitempty"`
-	Images     int `json:"images,omitempty"`
-	Projects   int `json:"projects,omitempty"`
-
-	VolumeStorage resource.Quantity `json:"volumeStorage,omitempty"`
-	Memory        resource.Quantity `json:"memory,omitempty"`
-	CPU           resource.Quantity `json:"cpu,omitempty"`
+	Flags            Flags            `json:"flags,omitempty"`
+	Counts           Counts           `json:"counts,omitempty"`
+	PersistentCounts PersistentCounts `json:"persistentCounts,omitempty"`
+	Quantities       Quantities       `json:"quantities,omitempty"`
+}
+
+// NewResources initializes and returns a new Resources instance with all maps initialized.
+func NewResources() Resources {
+	return Resources{
+		Flags:            make(Flags),
+		Counts:           make(Counts),
+		PersistentCounts: make(PersistentCounts),
+		Quantities:       make(Quantities),
+	}
 }
 
 // Add will add the resources of another Resources struct into the current one.
 func (current *Resources) Add(incoming Resources) {
-	current.Apps += incoming.Apps
-	current.Containers += incoming.Containers
-	current.Jobs += incoming.Jobs
-	current.Volumes += incoming.Volumes
-	current.Secrets += incoming.Secrets
-	current.Images += incoming.Images
-	current.Projects += incoming.Projects
-
-	current.VolumeStorage.Add(incoming.VolumeStorage)
-	current.Memory.Add(incoming.Memory)
-	current.CPU.Add(incoming.CPU)
+	// Add the incoming resources to the current ones
+	for flag, value := range incoming.Flags {
+		current.Flags[flag] = value
+	}
+	for count, value := range incoming.Counts {
+		current.Counts[count] += value
+	}
+	for quantity, value := range incoming.Quantities {
+		q := current.Quantities[quantity]
+		q.Add(value)
+	}
+	for persistentCount, value := range incoming.PersistentCounts {
+		current.PersistentCounts[persistentCount] += value
+	}
 }
 
 // Remove will remove the resources of another Resources struct from the current one.
 func (current *Resources) Remove(incoming Resources, all bool) {
-	// Do not allow resources to go below 0
-	nonNegativeSubtract := func(currentVal, incomingVal int) int {
-		difference := currentVal - incomingVal
-		if difference < 0 {
-			difference = 0
+	nonZeroSubtract := func(a, b int) int {
```

Comment:
I like the improvement to this function but I think the previous name made more sense:

```suggestion
	nonNegativeSubtract := func(a, b int) int {
```

Reply:
Yeah, just a typo. Good catch.